net/tcp/tcp_send_buffered.c

```diff
@@ -1172,7 +1204,12 @@ ssize_t psock_tcp_send(FAR struct socket *psock, FAR const void *buf,
               goto errout_with_lock;
             }
 
-          net_lockedwait_uninterruptible(&conn->snd_sem);
+          remain = tcp_send_gettimeout(anchor, remain);
+          ret = net_timedwait_uninterruptible(&conn->snd_sem, remain);
```

Comment:
Is it possible to go with
```
  clock_t start;
...
  start = clock_systime_ticks();
...
  ret = net_timedwait_uninterruptible(&conn->snd_sem, tcp_send_gettimeout(start, _SO_TIMEOUT(conn->sconn.s_sndtimeo)));
```
and modify `tcp_send_gettimeout` to
```
static unsigned int tcp_send_gettimeout(clock_t start, unsigned int timeout)
{
  clock_t elapse;

  if (remain != UINT_MAX)
    {
      elapse = TICK2MSEC(clock_systime_ticks() - start);
      if (elapse >= remain)
        {
          remain = 0;
        }
      else
        {
          remain -= elapse;
        }
    }

  return remain;
}
```
?

Reply:
1. anchor time should not to be changed on sending process, 
2. remain, anchor, _SO_TIMEOUT(conn->sconn.s_sndtimeo),  these are in milliseconds, which will reduce unnecessary type cast 
3. what different with current tcp_send_gettimeout ?

Reply:
> 1. anchor time should not to be changed on sending process,
> 

I do not propose to change anchor time on sending process. It will be done once. I just used a different name `anchor` -> `start`.
>     2. remain, anchor, _SO_TIMEOUT(conn->sconn.s_sndtimeo),  these are in milliseconds, which will reduce unnecessary type cast
>

I do not fully understand this. In my proposal I don't think that there are additional type casts.
>     3. what different with current tcp_send_gettimeout ?
In general I propose to track running time in `clock_t` (the same type that is returned by `clock_systime_ticks`) and perform conversion to `ms` only inside the `tcp_send_gettimeout` once. This will give better granularity compared to current code. Also it will eliminate having `uint32_t remain;` inside the `psock_tcp_send`.

Reply:
I can keep the anchor prototype as clock_t, but the remain still needs.
remain timeout parameter needs to be calculated dynamically, this time needs to be accumulated

```
 1207           remain = tcp_send_gettimeout(anchor, remain);
 1208           ret = net_timedwait_uninterruptible(&conn->snd_sem, remain);
                                                                //remain changed
1253               remain = tcp_send_gettimeout(anchor, remain);
1254               wrb = tcp_wrbuffer_timedalloc(remain);
1255               ninfo("new wrb %p\n", wrb);
1256               DEBUGASSERT(TCP_WBPKTLEN(wrb) == 0);
...
```

Reply:
I feel weird that my original commit was using clock_t, I change the clock_t to uint32_t cause of your comment

Reply:
> I can keep the anchor prototype as clock_t, but the remain still needs.
> remain timeout parameter needs to be calculated dynamically, this time needs to be accumulated
> 
> ```
>  1207           remain = tcp_send_gettimeout(anchor, remain);
>  1208           ret = net_timedwait_uninterruptible(&conn->snd_sem, remain);
>                                                                 //remain changed
> 1253               remain = tcp_send_gettimeout(anchor, remain);
> 1254               wrb = tcp_wrbuffer_timedalloc(remain);
> 1255               ninfo("new wrb %p\n", wrb);
> 1256               DEBUGASSERT(TCP_WBPKTLEN(wrb) == 0);
> ...
> ```

Please examine my suggestion again. It calculates parameter dynamically

Reply:
Done